pkg/backends/newrelic/newrelic.go

```diff
@@ -0,0 +1,441 @@
+		return nil, fmt.Errorf("[%s] metricsPerBatch must be positive", BackendName)
+	}
+	if clientTimeout <= 0 {
+		return nil, fmt.Errorf("[%s] clientTimeout must be positive", BackendName)
+	}
+	if maxRequestElapsedTime <= 0 {
+		return nil, fmt.Errorf("[%s] maxRequestElapsedTime must be positive", BackendName)
+	}
+
+	log.Infof("[%s] maxRequestElapsedTime=%s maxRequests=%d clientTimeout=%s metricsPerBatch=%d", BackendName, maxRequestElapsedTime, maxRequests, clientTimeout, metricsPerBatch)
+
+	dialer := &net.Dialer{
+		Timeout:   5 * time.Second,
+		KeepAlive: 30 * time.Second,
+	}
+	transport := &http.Transport{
+		Proxy:               http.ProxyFromEnvironment,
+		TLSHandshakeTimeout: 3 * time.Second,
+		TLSClientConfig: &tls.Config{
+			// Can't use SSLv3 because of POODLE and BEAST
+			// Can't use TLSv1.0 because of POODLE and BEAST using CBC cipher
+			// Can't use TLSv1.1 because of RC4 cipher usage
+			MinVersion: tls.VersionTLS12,
+		},
+		DialContext: func(ctx context.Context, _, address string) (net.Conn, error) {
+			// replace the network with our own
+			return dialer.DialContext(ctx, network, address)
+		},
+		MaxIdleConns:    50,
+		IdleConnTimeout: 1 * time.Minute,
+	}
+	if !enableHttp2 {
+		// A non-nil empty map used in TLSNextProto to disable HTTP/2 support in client.
+		// https://golang.org/doc/go1.6#http2
+		transport.TLSNextProto = map[string](func(string, *tls.Conn) http.RoundTripper){}
+	}
+
+	metricsBufferSem := make(chan *bytes.Buffer, maxRequests)
+	for i := uint(0); i < maxRequests; i++ {
+		metricsBufferSem <- &bytes.Buffer{}
+	}
+	eventsBufferSem := make(chan *bytes.Buffer, maxConcurrentEvents)
+	for i := uint(0); i < maxConcurrentEvents; i++ {
+		eventsBufferSem <- &bytes.Buffer{}
+	}
+	return &Client{
+		address:               address,
+		eventType:             eventType,
+		flushType:             flushType,
+		metricPrefix:          metricPrefix,
+		tagPrefix:             tagPrefix,
+		metricName:            metricName,
+		metricType:            metricType,
+		metricPerSecond:       metricPerSecond,
+		metricValue:           metricValue,
+		timerMin:              timerMin,
+		timerMax:              timerMax,
+		timerCount:            timerCount,
+		timerMean:             timerMean,
+		timerMedian:           timerMedian,
+		timerStdDev:           timerStdDev,
+		timerSum:              timerSum,
+		timerSumSquares:       timerSumSquares,
+		userAgent:             userAgent,
+		maxRequestElapsedTime: maxRequestElapsedTime,
+		client: http.Client{
+			Transport: transport,
+			Timeout:   clientTimeout,
+		},
+		metricsPerBatch:  metricsPerBatch,
+		metricsBufferSem: metricsBufferSem,
+		eventsBufferSem:  eventsBufferSem,
+		now:              time.Now,
+		flushInterval:    flushInterval,
+		disabledSubtypes: disabled,
+	}, nil
+}
+
+func getSubViper(v *viper.Viper, key string) *viper.Viper {
+	n := v.Sub(key)
+	if n == nil {
+		n = viper.New()
+	}
+	return n
+}
+
+func setDefaultPayload() NewRelicPayload {
```

Comment:
Updated this to newInfraPayload(). Reason is because it is a payload structure specific to the infrastructure agent. In future, I want to add another option to flush to our insights collector directly, where the payload is a slightly different so avoided the generic naming to future proof.